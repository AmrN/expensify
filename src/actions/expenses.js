// import uuid from 'uuid';
import database from '../firebase/firebase';

export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense,
});

export const startAddExpense = (expenseData = {}) => (dispatch) => {
  const {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0,
  } = expenseData;

  const expense = {
    description, note, amount, createdAt,
  };

  return database.ref('expenses').push(expense).then((ref) => {
    dispatch(addExpense({
      id: ref.key,
      ...expense,
    }));
  });
};

export const removeExpense = id => ({
  type: 'REMOVE_EXPENSE',
  id,
});

export const editExpense = (id, expense) => ({
  type: 'EDIT_EXPENSE',
  id,
  expense,
});


export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses,
});

export const startSetExpenses = () => dispatch => database.ref('expenses')
  .once('value')
  .then((snapshot) => {
    let expenses = snapshot.val() || {};
    expenses = Object.keys(expenses).map(id => ({ id, ...expenses[id] }));
    return dispatch(setExpenses(expenses));
  });

export const startRemoveExpense = id => dispatch => database.ref(`expenses/${id}`)
  .remove()
  .then(() => {
    dispatch(removeExpense(id));
  })
  .catch(e => console.log('Removing expense faild: ', e));

