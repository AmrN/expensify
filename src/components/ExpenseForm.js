import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    const { expense = {} } = props;
    this.state = {
      description: expense.description ? expense.description : '',
      amount: expense.amount ? (expense.amount / 100).toString() : '',
      note: expense.note ? expense.note : '',
      createdAt: expense.createdAt ? moment(expense.createdAt) : moment(),
      focusedCalendarInput: false,
      error: null,
    };
  }


  handleDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({
      description,
    }));
  }

  handleAmountChange = (e) => {
    const amount = e.target.value.trim();
    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount,
      }));
    }
  }

  handleNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({
      note,
    }));
  }

  handleSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount',
      }));
    } else {
      this.setState(() => ({
        error: null,
      }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  }

  handleDateChange = (date) => {
    date && this.setState({ createdAt: date });
  }

  handleFocusChange = ({ focused }) => {
    this.setState({
      focusedCalendarInput: focused,
    });
  }

  render() {
    return (
      <form
        onSubmit={this.handleSubmit}
        className="form"
      >
        {this.state.error &&
          <p className="form__error">
            {this.state.error}
          </p>
        }

        <input
          type="text"
          placeholder="Description"
          value={this.state.description}
          onChange={this.handleDescriptionChange}
          autoFocus
          className="text-input"
        />
        <input
          type="text"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.handleAmountChange}
          className="text-input"
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.handleDateChange}
          focused={this.state.focusedCalendarInput}
          onFocusChange={this.handleFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.handleNoteChange}
          className="textarea"
        />
        <button className="button">
          Save Expense
        </button>
      </form>
    );
  }
}

export default ExpenseForm;
