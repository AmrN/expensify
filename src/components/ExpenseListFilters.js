import React from 'react';
import { connect } from 'react-redux';
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
} from '../actions/filters';
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';


export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
  };

  handleDatesChange = ({ startDate, endDate }) => {
    const { setStartDate, setEndDate } = this.props;
    setStartDate(startDate);
    setEndDate(endDate);
  };

  handleFocusChange = (calendarFocused) => {
    this.setState(() => ({
      calendarFocused,
    }))
  }

  handleTextChange = (e) => {
    this.props.setTextFilter(e.target.value)
  }

  handleSortChange = (e) => {
    e.target.value === 'createdAt' ? this.props.sortByDate() : this.props.sortByAmount()
  }

  render() {
    const {
      filters,
      setTextFilter,
      sortByAmount,
      sortByDate,
     } = this.props;

    return (
      <div>
        <input
          type="text"
          value={filters.text}
          onChange={this.handleTextChange} />

        <select
          value={filters.sortBy}
          onChange={this.handleSortChange}
        >
          <option value="createdAt">Date</option>
          <option value="amount">Amount</option>
        </select>

        <DateRangePicker
          startDate={filters.startDate}
          endDate={filters.endDate}
          onDatesChange={this.handleDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.handleFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters,
});

export default connect(mapStateToProps, {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
})(ExpenseListFilters);