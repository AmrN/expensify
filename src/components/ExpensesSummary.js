import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
  const expensesWord = `expense${expensesCount === 1 ? '' : 's'}`;
  const expensesTotalFormatted = numeral(expensesTotal / 100).format('$0,0.00');

  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{expensesCount}</span> {expensesWord} totalling <span>{expensesTotalFormatted}</span>
        </h1>
        <div className="page-header__actions">
          <Link
            to="/create"
            className="button"
          >
            Add Expense
          </Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
  return {
    expensesCount: visibleExpenses.length,
    expensesTotal: getExpensesTotal(visibleExpenses),
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
