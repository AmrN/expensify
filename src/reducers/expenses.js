const expensesReducer = (state = [], action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense,
      ];
    case 'REMOVE_EXPENSE':
      return state.filter(el => el.id !== action.id);
    // return [
    //   ...state.slice(0, itemIndex),
    //   ...state.slice(itemIndex + 1),
    // ];
    case 'EDIT_EXPENSE':
      return state.map(el => (el.id === action.id ? { ...el, ...action.expense } : el));
    // return [
    //   ...state.slice(0, expenseIndex),
    //   {...state[expenseIndex], ...action.expense},
    //   ...state.slice(expenseIndex+1),
    // ];
    case 'SET_EXPENSES':
      return [...action.expenses];
    default:
      return state;
  }
};

export default expensesReducer;
