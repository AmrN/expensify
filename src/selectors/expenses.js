import moment from 'moment';

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  const res = expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(expense.createdAt, 'day') : true,
      endDateMatch = endDate ? endDate.isSameOrAfter(expense.createdAt, 'day') : true,
      textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  });
  return res.sort((a, b) => a[sortBy] > b[sortBy] ? -1 : 1)
};

export default getVisibleExpenses;