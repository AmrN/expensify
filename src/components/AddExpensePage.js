import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  handleSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={this.handleSubmit}
        />
      </div>
    );
  }
}

export default connect(null, { startAddExpense })(AddExpensePage);
