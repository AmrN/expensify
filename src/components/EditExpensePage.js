import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  handleExpenseFormSubmit = (updatedExpense) => {
    this.props.editExpense(updatedExpense.id, updatedExpense);
    this.props.history.push('/');
  }

  handleRemoveExpense = () => {
    this.props.removeExpense(this.props.expense.id);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.handleExpenseFormSubmit}
        />
        <button onClick={this.handleRemoveExpense}>remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  expense: state.expenses.find((el) => el.id === ownProps.match.params.id),
});

export default connect(mapStateToProps, { editExpense, removeExpense })(EditExpensePage);