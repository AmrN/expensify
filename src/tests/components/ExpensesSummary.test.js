import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render ExpensesSummary correctly when passing multiple expenses', () => {
  const wrapper = shallow(
    <ExpensesSummary
      expensesCount={3}
      expensesTotal={3000} />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary correctly when passing a single expense', () => {
  const wrapper = shallow(
    <ExpensesSummary
      expensesCount={1}
      expensesTotal={1000} />
  );
  expect(wrapper).toMatchSnapshot();
});