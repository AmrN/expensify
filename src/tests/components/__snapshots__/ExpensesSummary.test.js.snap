// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly when passing a single expense 1`] = `
<h3>
  Viewing 
  1
   expense
   totalling 
  $10.00
</h3>
`;

exports[`should render ExpensesSummary correctly when passing multiple expenses 1`] = `
<h3>
  Viewing 
  3
   expense
  s
   totalling 
  $30.00
</h3>
`;
